src/proto/single/ent.rs

```diff
@@ -74,6 +91,135 @@ impl JobBuilder {
     }
 }
 
+/// Info on job execution progress (retrieved).
+///
+/// The tracker is guaranteed to get the following details: the job's id (though
+/// they should know it beforehand in order to be ably to track the job), its last
+/// know state (e.g."enqueued", "working", "success", "unknown") and the date and time
+/// the job was last updated. Additionally, information on what's going on with the job
+/// ([desc](struct.ProgressUpdate.html#structfield.desc)) and completion percentage
+/// ([percent](struct.ProgressUpdate.html#structfield.percent)) may be available,
+/// if the worker provided those details.
+#[derive(Debug, Clone, Deserialize)]
+pub struct Progress {
+    /// Id of the tracked job.
+    pub jid: String,
+
+    /// Job's state.
+    pub state: String,
+
+    /// When this job was last updated.
+    #[serde(deserialize_with = "parse_datetime")]
+    pub updated_at: Option<DateTime<Utc>>,
+
+    /// Percentage of the job's completion.
+    pub percent: Option<u8>,
+
+    /// Arbitrary description that may be useful to whoever is tracking the job's progress.
+    pub desc: Option<String>,
+}
+
+/// Info on job execution progress (sent).
+///
+/// In Enterprise Faktory, a client executing a job can report on the execution
+/// progress, provided the job is trackable. A trackable job is the one with "track":1
+/// specified in the custom data hash.
+#[derive(Debug, Clone, Serialize, Builder)]
+#[builder(
+    custom_constructor,
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct ProgressUpdate {
+    /// Id of the tracked job.
+    #[builder(setter(custom))]
+    pub jid: String,
+
+    /// Percentage of the job's completion.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub percent: Option<u8>,
+
+    /// Arbitrary description that may be useful to whoever is tracking the job's progress.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub desc: Option<String>,
+
+    /// Allows to extend the job's reservation, if more time needed to execute it.
+    ///
+    /// Note that you cannot decrease the initial [reservation](struct.Job.html#structfield.reserve_for).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub reserve_until: Option<DateTime<Utc>>,
+}
+
+impl ProgressUpdate {
+    /// Create a new instance of `ProgressUpdateBuilder` with job ID already set.
+    ///
+    /// Equivalent to creating a [new](struct.ProgressUpdateBuilder.html#method.new)
+    /// `ProgressUpdateBuilder`.
+    pub fn builder(jid: impl Into<String>) -> ProgressUpdateBuilder {
+        ProgressUpdateBuilder {
+            jid: Some(jid.into()),
+            ..ProgressUpdateBuilder::create_empty()
```

Comment:
I assume `create_empty` isn't `pub`?

Reply:
It is not. But those are generated methods and worth double-checking.
![image](https://github.com/jonhoo/faktory-rs/assets/117771945/262e2122-b587-461d-88da-1831b6daa0a5)